lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,79 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+IdentifierInfo *Parser::getBoundsExpressionStart() {
+  if (Tok.getKind() == tok::identifier) {
+    IdentifierInfo *Ident = Tok.getIdentifierInfo();
+    if (Ident == Ident_bounds || Ident == Ident_byte_count || Ident == Ident_count) {
+      return Ident;
+    }
+  }
+  return nullptr;
+}
+
+ExprResult Parser::ParseBoundsExpression() {
+  IdentifierInfo *Ident = getBoundsExpressionStart();
+  if (!Ident) {
+    Diag(Tok, diag::err_expected_bounds_expr);
+    return ExprError();
+  }
+
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+  if (ExpectAndConsume(tok::l_paren)) {
+    return ExprError();
+  }
+
+  ExprResult Result;
+  if (Ident == Ident_byte_count || Ident == Ident_count) {
+    Result = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
+    CountBoundsExpr::Kind CountKind = Ident == Ident_count ?
+      CountBoundsExpr::Kind::Count : CountBoundsExpr::Kind::Byte_Count;
+    if (!Result.isInvalid()) {
+      Result = Actions.ActOnCountBoundsExpr(BoundsKWLoc, CountKind, Result.get(),
+                                           Tok.getLocation());
+    } else {
+      Result = ExprError();
+    }
+  } else if (Ident == Ident_bounds) {
+    bool FoundNullaryOperator = false;
+
+    if (Tok.getKind() == tok::identifier) {
+      IdentifierInfo *NextIdent = Tok.getIdentifierInfo();
+      if (NextIdent == Ident_none) {
```

Comment:
None is a contextual keyword.   You can't write bounds(none, y).  But you could write bounds(y, none), and none will be treated as a variable name.